### ISSUE TITLE
Jerky behavior when tracking with Gemini

### ISSUE THREAD

**Issue by bgottula** [bgottula/track#79]
When testing hybrid tracking I observed jerky behavior in the mount while it was slewing to bright stars. I have seen this behavior early in my experimentation with the Losmandy mount but I can no longer remember the source of the trouble. The jerkiness becomes more pronounced as the mount converges on the target, when the slew rates have fallen below the max limit. What I do recall is that something was causing occasional spikes in the error signal, either the mount's reported position or the target position. Reproduce if possible and figure out what the root cause is.

**Comment by bgottula** [comment]
Not sure if this is the exact same issue but I encountered a similar problem on 10 April while attempting to track a satellite pass with hybrid mode. I found that the mount would jerk and stop repeatedly, and the display showed DEC motor stalls. After a couple attempts I switched to blind mode which worked perfectly, so this seems to be an issue specific to hybrid mode. One possibility is that the acceleration limit does not work well when the delay between loop filter updates is long, which could be caused by the camera framerate or the amount of time it takes for the camera to initialize.

**Comment by bgottula** [comment]
I did a couple of experiments which shed some light on the problem. First I ran hybrid_track from the home position to the star Caph (with no alignment performed) and confirmed that the jerky behavior was present. Along with the jerkiness the point package was printing messages indicating that it was dealing with NACKs. I then added some print statements to the control loop along with the commanded slew rates at each iteration to see if the acceleration limits were being enforced. Based on this data it appears that the acceleration limit (10 degrees/s^2) is being enforced properly. However it is also evident that some of the control loop periods are quite lengthy. Some of this is possibly due to the NACK timeout being 0.25 seconds by default but it is also known that the camera frame rate is fairly slow.

```
slew rates: {'dec': 0.0, 'ra': 0.0}
slew rates: {'dec': 0.6241774559020996, 'ra': -0.6308531761169434}
slew rates: {'dec': 1.2669920921325684, 'ra': -1.2623071670532227}
slew rates: {'dec': 2.2200775146484375, 'ra': -2.215735912322998}
after 1 NACK's, Gemini indicated that our command datagram was lost; will resend it
Warning: loop filter update period was 0.3142871856689453 s, limit is 0.1 s.
slew rates: {'dec': 1.495930777966182, 'ra': -2.8546571731567383}
after 1 NACK's, Gemini indicated that our command datagram was lost; will resend it
Warning: loop filter update period was 0.5370352268218994 s, limit is 0.1 s.
Warning: loop filter update period was 0.2862091064453125 s, limit is 0.1 s.
slew rates: {'dec': 1.495930777966182, 'ra': -2.8546571731567383}
```

I repeated the same exercise with the blind_track program and confirmed again that it does not experience the same issue. The control loop is still riding the acceleration limit, but the big difference is that the control loop period is much shorter. Therefore the delta between each commanded rate is also much smaller.

```
time: 1525232197.3367388 slew rates: {'dec': 0.0, 'ra': 0.0}
time: 1525232197.3510845 slew rates: {'dec': 0.14302730560302734, 'ra': -0.14398574829101562}
time: 1525232197.365412 slew rates: {'dec': 0.2866053581237793, 'ra': -0.287628173828125}
time: 1525232197.3797526 slew rates: {'dec': 0.42991161346435547, 'ra': -0.4315352439880371}
time: 1525232197.39409 slew rates: {'dec': 0.5733728408813477, 'ra': -0.5750727653503418}
time: 1525232197.4085026 slew rates: {'dec': 0.7164764404296875, 'ra': -0.7187294960021973}
time: 1525232197.4228356 slew rates: {'dec': 0.8606266975402832, 'ra': -0.863187313079834}
time: 1525232197.4371505 slew rates: {'dec': 1.0038352012634277, 'ra': -1.0062575340270996}
time: 1525232197.4515104 slew rates: {'dec': 1.1474251747131348, 'ra': -1.149604320526123}
time: 1525232197.4658573 slew rates: {'dec': 1.2908554077148438, 'ra': -1.2929797172546387}
time: 1525232197.480145 slew rates: {'dec': 1.4338397979736328, 'ra': -1.436452865600586}
time: 1525232197.4944694 slew rates: {'dec': 1.5767955780029297, 'ra': -1.5793299674987793}
time: 1525232197.5088024 slew rates: {'dec': 1.720283031463623, 'ra': -1.7225146293640137}
time: 1525232197.5235279 slew rates: {'dec': 1.8674850463867188, 'ra': -1.8660449981689453}
time: 1525232197.5378406 slew rates: {'dec': 2.0107340812683105, 'ra': -2.0132994651794434}
time: 1525232197.5509586 slew rates: {'dec': 2.15378999710083, 'ra': -2.156364917755127}
time: 1525232197.5651338 slew rates: {'dec': 2.2837376594543457, 'ra': -2.2864484786987305}
time: 1525232197.579471 slew rates: {'dec': 2.4269700050354004, 'ra': -2.4289870262145996}
time: 1525232197.5938458 slew rates: {'dec': 2.5707173347473145, 'ra': -2.5729870796203613}
time: 1525232197.6082473 slew rates: {'dec': 2.714824676513672, 'ra': -2.716848850250244}
time: 1525232197.6219628 slew rates: {'dec': 2.851865291595459, 'ra': -2.860691547393799}
time: 1525232197.6363106 slew rates: {'dec': 2.995433807373047, 'ra': -2.997894287109375}
time: 1525232197.6500182 slew rates: {'dec': 3.133528232574463, 'ra': -3.141498565673828}
time: 1525232197.6643765 slew rates: {'dec': 3.2759952545166016, 'ra': -3.2783150672912598}
time: 1525232197.678769 slew rates: {'dec': 3.419828414916992, 'ra': -3.4219861030578613}
time: 1525232197.6931515 slew rates: {'dec': 3.5638070106506348, 'ra': -3.565850257873535}
time: 1525232197.7075486 slew rates: {'dec': 3.707599639892578, 'ra': -3.7097620964050293}
time: 1525232197.7219281 slew rates: {'dec': 3.8515114784240723, 'ra': -3.8536643981933594}
time: 1525232197.7363238 slew rates: {'dec': 3.995664119720459, 'ra': -3.997485637664795}
time: 1525232197.7507634 slew rates: {'dec': 4.0, 'ra': -4.0}
```

At this point I suspected that I could elicit the same behavior from the blind_track program by artificially increasing the loop period. I inserted a sleep statement in control.py to ensure that each iteration was at least 0.1 seconds long. This reproduced the stalls and NACKs.

```
time: 1525232310.4415672 slew rates: {'ra': 0.0, 'dec': 0.0}
Warning: loop filter update period was 0.10100317001342773 s, limit is 0.1 s.
Warning: loop filter update period was 0.10140109062194824 s, limit is 0.1 s.
time: 1525232310.5429397 slew rates: {'ra': 0.0, 'dec': 0.0}
Warning: loop filter update period was 0.10042667388916016 s, limit is 0.1 s.
Warning: loop filter update period was 0.10048103332519531 s, limit is 0.1 s.
time: 1525232310.6434164 slew rates: {'ra': 0.0, 'dec': 0.0}
time: 1525232310.742162 slew rates: {'ra': -0.9871172904968262, 'dec': 0.9873199462890625}
Warning: loop filter update period was 0.1012260913848877 s, limit is 0.1 s.
Warning: loop filter update period was 0.10116004943847656 s, limit is 0.1 s.
time: 1525232310.843336 slew rates: {'ra': -0.9871172904968262, 'dec': 0.6101196543583891}
time: 1525232310.9423337 slew rates: {'ra': -1.977541446685791, 'dec': 1.9771647453308105}
Warning: loop filter update period was 0.3013741970062256 s, limit is 0.1 s.
Warning: loop filter update period was 0.30147767066955566 s, limit is 0.1 s.
time: 1525232311.2437975 slew rates: {'ra': -1.977541446685791, 'dec': 1.2184566881496592}
after 1 NACK's, Gemini indicated that our command datagram was lost; will resend it
Warning: loop filter update period was 0.28010988235473633 s, limit is 0.1 s.
time: 1525232311.5238025 slew rates: {'ra': -2.1215224266052246, 'dec': 1.2184566881496592}
Warning: loop filter update period was 0.27978014945983887 s, limit is 0.1 s.
after 1 NACK's, Gemini indicated that our command datagram was lost; will resend it
Warning: loop filter update period was 0.28010129928588867 s, limit is 0.1 s.
time: 1525232311.8039074 slew rates: {'ra': -2.1215224266052246, 'dec': 1.2184566881496592}
Warning: loop filter update period was 0.2800867557525635 s, limit is 0.1 s.
```

Next I decided to try reducing the acceleration limit to see if that would help. I reduced it to 1 degree/s^2. This did help, but the acceleration was much too slow and the motion of the mount was still very jerky. In fact in some ways it was worse because the acceleration limit was now interfering with the control loop response.

It seems like the ideal solution to the problem would be to use threads or multiprocessing to allow slew commands to the mount to be issued at a higher frequency than the camera framerate. That would allow smooth but rapid accelerations. The only simple alternative I can think of is imposing an additional limit on the maximum magnitude change in slew rate between consecutive loop iterations. This would be in addition to the existing acceleration limit. I'm not super optimistic that this will have the desired effect though.

**Comment by bgottula** [comment]
Enforcing a maximum slew rate step size of 0.1 degrees/s was effective at preventing stalls but it did result in slower than desired acceleration and some underdamped behavior when the mount reached the target. 0.2 degrees/s also seemed to be okay with better responsiveness. At 0.3 the mount only overshot the target once in each axis and the motors still were not stalling. At 0.4 responsiveness was still better. 0.5 was better still. I then jumped up to a step limit of 1.0 and this was too high: motor stalls occurred. 0.8 was also too jerky, though I'm not certain that a stall occurred in this case. 0.6 also seemed too harsh.

Overall I felt that 0.4 degrees/s was the sweet spot. I have a feeling this result is dependent on the total mass of the mount's payload. For this testing I had the 130SLT installed.

**Comment by bgottula** [comment]
Based on the success of the step limit testing I will implement that fix for now (and perhaps forever). Smoother operation would be possible by use of threading but this introduces a great deal of extra complexity. We have already been down that road a couple of times and usually regretted it.

**Comment by bgottula** [comment]
Reopening because while testing the align.py script indoors I observed some jerkiness in the mount's movement when the mount was converging on the target and the slew rate was fairly slow. The jerks seemed to correlate with warnings printed to the console indicating that the loop period exceeded 0.1 seconds.

**Comment by bgottula** [comment]
A likely solution is to modify the hybrid error source such that when in blind mode it does not block on trying to get a frame from the camera. The benefit of this approach is that blind tracking can always run with a fast control cycle whether using the blind error source or the hybrid error source. This was tested successfully during a tracking session using nix but the code has not been committed yet.

**Comment by bgottula** [comment]
Committed the code to allow non-blocking camera operation when hybrid error source is in blind mode. Attempted to test but was hampered ironically by excessive motor stalls. Not sure what the problem is, but it was occurring even with the blind_track program which is unaffected by these modifications. Perhaps the problem is that the mount is currently not loaded with an OTA or counterweights.

**Comment by bgottula** [comment]
The jerky behavior I observed before I wrote the previous comment was not caused by mount loading or balance after all. I have no idea what the issue was, but the RA axis was stuttering like crazy but only when motion was in one direction but not the other. The behavior was present when using the gamepad to control it as well. I don't know what I did to recover, but somehow after parking at CWD position I am now able to blind slew to any object and the mount motion is super smooth and stall/jerk free. Very strange. I wonder if there are bugs in the Gemini firmware that causes this.
